lib/fileset/default.nix

```
@@ -123,9 +172,113 @@ The only way to change which files get added to the store is by changing the `fi
+  The file set containing all files that are in either of two given file sets.
+  This is the same as [`unions`](#function-library-lib.fileset.unions),
+  but takes just two file sets instead of a list.
+  See also [Union (set theory)](https://en.wikipedia.org/wiki/Union_(set_theory)).
+
+  The given file sets are evaluated as lazily as possible,
+  with the first argument being evaluated first if needed.
+
+  Type:
+    union :: FileSet -> FileSet -> FileSet
+
+  Example:
+    # Create a file set containing the file `Makefile`
+    # and all files recursively in the `src` directory
+    union ./Makefile ./src
+
+    # Create a file set containing the file `Makefile`
+    # and the LICENSE file from the parent directory
+    union ./Makefile ../LICENSE
+  */
+  union =
+    # The first file set.
+    # This argument can also be a path,
+    # which gets [implicitly coerced to a file set](#sec-fileset-path-coercion).
+    fileset1:
+    # The second file set.
+    # This argument can also be a path,
+    # which gets [implicitly coerced to a file set](#sec-fileset-path-coercion).
+    fileset2:
+    let
+      filesets = _coerceMany "lib.fileset.union" [
+        {
+          context = "first argument";
+          value = fileset1;
+        }
+        {
+          context = "second argument";
+          value = fileset2;
+        }
+      ];
+    in
+    _unionMany filesets;
+
+  /*
+  The file set containing all files that are in any of the given file sets.
+  This is the same as [`union`](#function-library-lib.fileset.unions),
+  but takes a list of file sets instead of just two.
+  See also [Union (set theory)](https://en.wikipedia.org/wiki/Union_(set_theory)).
+
+  The given file sets are evaluated as lazily as possible,
+  with earlier elements being evaluated first if needed.
+
+  Type:
+    unions :: [ FileSet ] -> FileSet
+
+  Example:
+    # Create a file set containing selected files
+    unions [
+      # Include the single file `Makefile` in the current directory
+      # This errors if the file doesn't exist
+      ./Makefile
+
+      # Recursively include all files in the `src/code` directory
+      # If this directory is empty this has no effect
+      ./src/code
+
+      # Include the files `run.sh` and `unit.c` from the `tests` directory
+      ./tests/run.sh
+      ./tests/unit.c
+
+      # Include the `LICENSE` file from the parent directory
+      ../LICENSE
+    ]
+  */
+  unions =
+    # A list of file sets.
+    # Must contain at least 1 element.
+    # The elements can also be paths,
+    # which get [implicitly coerced to file sets](#sec-fileset-path-coercion).
+    filesets:
+    let
+      # We cannot rename matched attribute arguments, so let's work around it with an extra `let in` statement
+      maybeFilesets = filesets;
+    in
+    let
+      # Annotate the elements with context, used by _coerceMany for better errors
+      annotated = imap0 (i: el: {
+        context = "element ${toString i} of the argument";
```

Not sure. Might this be cleaner?

```suggestion
        context = "element ${toString i} of the filesets.unions argument";
```

---

```suggestion
        context = "element ${toString i} of the argument to `fileset.unions`";
```

---

The error already shows `fileset.unions` in the context:

```
lib.fileset.unions: element 0 of the argument "test" is a string-like value, but it should be a path instead.
```

I guess the `argument "test"` is a bit confusing though, I'd be in favor of this instead:

```
lib.fileset.unions: element 0 ("/foo/bar") is a string-like value, but it should be a path instead.
```